### PR TITLE
Fix use of `PBENCH_REDIS_SERVER` by both `pbench-tool-meister-client` and `-stop`

### PR DESCRIPTION
**Pull request by portante**
When we implemented PR #2181, we failed to properly test for the bind/connect style of specification in our tests, and missed the fact that we didn't update `pbench-tool-meister-client` and `pbench-tool-meister-stop` to share the same code as is used by `pbench-tool-meister-start`.

There were 5 initial commits in this PR which should were squashed to one once they were reviewed:

 1. Refactor to share a `BaseReturnCode` class in `agent/utils.py`
 2. Refactor the `BaseServer` class into the `agent/utils.py` module
 3. Add tests that demonstrate things break
 4. Fix `pbench-tool-meister-stop`
 5. Fix `pbench-tool-meister-client`

This refactoring made it easier to see the progression of changes, facilitating a hopefully easier review.

The final review comment follows below.

----

Prior to this commit, `pbench-tool-meister-start` was the only CLI interface which was using the two forms for specifying the Redis server and Tool Data Sink host/port for binding and connecting.  Both the CLI commands for stopping the Tool Meister sub-system and communicating with it incorrectly expected the presence of the `PBENCH_REDIS_SERVER` to only specify a `<host>:<port>`, and also indicate that instance was not managed by CLI commands on this controller.

This commit corrects that where both `pbench-tool-meister-client` and `pbench-tool-meister-stop` now properly handle both forms of host/port specification, and `-stop` determines whether or not the CLI is in charge of orchestrating the server components by looking for the local pid files.

This was accomplished via a series of refactorings to share code across those three CLI commands:

   * Add and use `BaseReturnCode` in `lib/pbench/agent/utils.py`
   * Move BaseServer to `.../agent/utils.py`
   * Add a `RedisServer` class for `pbench-tool-meister-stop` to take advantage of the common parameter handling of the Redis server specification
   * In `pbench-tool-meister-client`, we use the new `RedisServerCommon` class shared with `-start` and `-stop`
   * We fix the new `BaseServer.kill()` to exit immediately after posting a `SIGKILL`
   * Add unit tests for all common code moved to `.../agent/utils.py`

### CODE REVIEW

File: lib/pbench/agent/utils.py
Comment (portante) [inline lib/pbench/agent/utils.py:32]: Looks like this can be removed.

**Review by portante — COMMENTED** [review]
Looks good for initial review.

File: agent/util-scripts/pbench-tool-meister-stop
Comment (dbutenhof) [inline agent/util-scripts/pbench-tool-meister-stop:67]: Why not just `return self.pid_file is not None`??

(You could just `return self.pid_file` but I do like the idea of converting the file path to a boolean here...)

**Review by dbutenhof — DISMISSED** [review]
I think this looks OK...

File: agent/util-scripts/pbench-tool-meister-stop
Comment (portante) [reply agent/util-scripts/pbench-tool-meister-stop:67]: Ah, yeah, good question.  I'll fix.

File: lib/pbench/agent/utils.py
Comment (ndokos) [inline lib/pbench/agent/utils.py:125]: Is it worth making this a function and adding an all-possible-permutations test for it?

**Review by ndokos — CHANGES_REQUESTED** [review]

File: agent/util-scripts/pbench-tool-meister-stop
Comment (webbnh) [reply agent/util-scripts/pbench-tool-meister-stop:67]: Extending @dbutenhof's suggestion:
```
    def locally_managed(self):
        return bool(self.pid_file)
```

:-)

**Review by webbnh — APPROVED** [review]
Looks OK.  (In fact, there are some nice improvements there!)

I'm not crazy about the parallel definitions of `RedisServer`, but I can guess why you did that.  It seems like it would be better to provide a single implementation which includes the capabilities which each consumer needs, although in at least one case it seems like you could get by on the base class alone.

And, of course, there is the usual collection of nits and suggestions.

**Comment by RobertKrawitz** [comment]
+1
Verified against bench-army-knife.

File: agent/util-scripts/pbench-tool-meister-stop
Comment (portante) [reply agent/util-scripts/pbench-tool-meister-stop:67]: Because this would return `True` for any `Path` object, even if that object is not associated with a path.

File: lib/pbench/agent/utils.py
Comment (portante) [reply lib/pbench/agent/utils.py:125]: Yes, good idea, I should add proper unit tests for this refactored code.

File: agent/util-scripts/pbench-tool-meister-start
Comment (portante) [reply agent/util-scripts/pbench-tool-meister-start:143]: > I'm unclear on how we're going to avoid name collisions on values

Good question, I'll investigate; one possibility is to employ a unit test to compare the base class codes against the sub-classes to ensure no overlap.

> how we avoid unintentional value collisions

The way the `ReturnCode` "kill" works is it shifts the codes by 1,000.

File: agent/util-scripts/pbench-tool-meister-stop
Comment (portante) [reply agent/util-scripts/pbench-tool-meister-stop:118]: The `kill()` method of `BaseServer` takes a return value to pass along on success, or change to a `KILL_` value.

File: agent/util-scripts/pbench-tool-meister-client
Comment (portante) [reply agent/util-scripts/pbench-tool-meister-client]: > Why isn't the definition of `RedisServer` common and shared with `pbench-tool-meister-start` and `-stop`?

Basically because it _seemed_ it would result in more code for very little gain.

At the time I did not want to embark on solving the return codes for `-stop` and `-client` while doing this work, but just wanted the shared constructor behavior.

Thank you for prompting me to look at this a bit more, as I think we can solve this elegantly.

File: agent/util-scripts/pbench-tool-meister-stop
Comment (webbnh) [reply agent/util-scripts/pbench-tool-meister-stop]: `bool(self.pid_file)` is roughly equivalent to `self.pid_file not in [0, False, None, "", (), [], {}]`, so it's actually a superset of `self.pid_file is not None`.

That is, if `self.pid_file` is a `Path` object (or anything else which is not `None`), your code will return `True` as well, just like mine does.

There's only a difference if `self.pid_file` somehow evaluates to a "Falsey" value other than `None`.

File: agent/util-scripts/pbench-tool-meister-start
Comment (webbnh) [reply agent/util-scripts/pbench-tool-meister-start:144]: OK, so the `KILL_` values are a different "type".  (One wonders if they shouldn't be in their own `IntEnum`.)

File: agent/util-scripts/pbench-tool-meister-stop
Comment (webbnh) [inline agent/util-scripts/pbench-tool-meister-stop:21]: `pbench-tool-meister-start` references `def_redis_port` in its help text.  Should `pbench-tool-meister-stop` be doing the same?  Or, better yet, should the two tools be sharing their argument definitions for the pieces which should be uniform between them?

File: agent/util-scripts/pbench-tool-meister-stop
Comment (webbnh) [inline agent/util-scripts/pbench-tool-meister-stop:74]: There's no description (nor type hint) for the return value from this `graceful_shutdown()` (nor for `locally_managed()`, but that one is easier to infer).

**Review by webbnh — APPROVED** [review]
Looks good (the refactor of `RedisServerCommon` is a solid improvement, although whether the server is locally managed or not seems like it would be of broader interest, but it's OK as is).

Two items for your consideration.

File: lib/pbench/agent/utils.py
Comment (portante) [reply lib/pbench/agent/utils.py:128]: Done.

File: agent/util-scripts/pbench-tool-meister-start
Comment (portante) [reply agent/util-scripts/pbench-tool-meister-start:144]: Tracking in #2235.

**Review by portante — COMMENTED** [review]
Okay, added a new commit with the unit tests.  Will rebase before merging.

**Review by portante — COMMENTED** [review]
Okay, missed a few things, and fixed LGTM comment ...

File: agent/util-scripts/pbench-tool-meister-start
Comment (ndokos) [inline agent/util-scripts/pbench-tool-meister-start:215]: No type hint here? Isn't it a `RedisServerCommon`? Or maybe a `RedisServer`? Is it n ordering problem (the class is not defined yet)?

File: agent/util-scripts/pbench-tool-meister-start
Comment (ndokos) [inline agent/util-scripts/pbench-tool-meister-start:353]: Type hint?

**Review by ndokos — COMMENTED** [review]
A couple of comments, plus I still need to review `test_utils.py`.

File: agent/util-scripts/pbench-tool-meister-start
Comment (portante) [reply agent/util-scripts/pbench-tool-meister-start:215]: Yes, ordering problem and I did not want to blow up the code with a major code move as part of the PR.

File: agent/util-scripts/pbench-tool-meister-start
Comment (portante) [reply agent/util-scripts/pbench-tool-meister-start:353]: Yes, ordering problem and I did not want to blow up the code with a major code move as part of the PR.

File: lib/pbench/test/unit/agent/test_utils.py
Comment (webbnh) [inline lib/pbench/test/unit/agent/test_utils.py:32]: Shouldn't the RHS be `OurServer.def_port` rather than a copy of the magic number?

Likewise, shouldn't the RHS in line 41 be `OurServer.bad_host_ret_code`?

Etc.

Comment (webbnh) [reply lib/pbench/test/unit/agent/test_utils.py:32]: Or, if you think accessing the class attributes is too "risky", then it seems like you should be using symbolic constants for them defined earlier and used to initialize the class.

File: lib/pbench/test/unit/agent/test_utils.py
Comment (webbnh) [inline lib/pbench/test/unit/agent/test_utils.py:118]: That doesn't look much like an [`OSError`](https://docs.python.org/3/library/exceptions.html#OSError) -- it should have an `errno` code as it's first argument, I think.

Comment (webbnh) [reply lib/pbench/test/unit/agent/test_utils.py:118]: Rather than having the c'tor accept a string as an argument (type hints would be handy, BTW), how about having it accept the exception object to raise?  That would decouple and simplify things a bit.

File: lib/pbench/test/unit/agent/test_utils.py
Comment (webbnh) [inline lib/pbench/test/unit/agent/test_utils.py:148]: The third column ("killed by term"?) is always `False`.  Are you missing a test case, or should this column be omitted from the table?

**Review by webbnh — APPROVED** [review]
Looks OK...unless you want to address any of the items below.

File: lib/pbench/agent/utils.py
Comment (portante) [reply lib/pbench/agent/utils.py:155]: Of the code being changed in this commit, I have made this modification as a separate commit for those files that were touched.  I am not going to make this change for any other module.

File: agent/util-scripts/pbench-tool-meister-stop
Comment (portante) [reply agent/util-scripts/pbench-tool-meister-stop:81]: Fixed in a separate commit.

File: agent/util-scripts/pbench-tool-meister-start
Comment (portante) [reply agent/util-scripts/pbench-tool-meister-start]: Fixed in a separate commit.

File: lib/pbench/test/unit/agent/test_utils.py
Comment (portante) [reply lib/pbench/test/unit/agent/test_utils.py]: Fixed as a separate commit.

File: lib/pbench/test/unit/agent/test_utils.py
Comment (portante) [reply lib/pbench/test/unit/agent/test_utils.py:152]: The first case was incorrect, fixed in a separate commit.

File: lib/pbench/test/unit/agent/test_utils.py
Comment (portante) [reply lib/pbench/test/unit/agent/test_utils.py:122]: Fixed as a separate commit.

**Review by portante — COMMENTED** [review]
There are 16 commits in this PR, which we'll want to squash before merging.

I have set this to `Draft` mode until we deem the code fit, then I'll squash them all down and we can finalize the approvals.

File: lib/pbench/test/unit/agent/test_utils.py
Comment (webbnh) [inline lib/pbench/test/unit/agent/test_utils.py:35]: Since you made the port value dynamic, do you want to make the server name and/or the hostname dynamic?

Alternatively, if this is only used for debugging, do we really want to be testing it at all?

File: lib/pbench/test/unit/agent/test_utils.py
Comment (webbnh) [inline lib/pbench/test/unit/agent/test_utils.py:43]: I hate to ask, but...does `excinfo` have a `value` attribute?  If so, what is it?

(And, is `f"{excinfo.value!r}"` better than `repr(excinfo.value)`?)

Ditto line 49.

**Review by webbnh — APPROVED** [review]
If your intent is to squash this branch into a single commit on the end of the `main` branch, I believe that you can do so using the button below after you get the approvals, without having to do an external squash and the subsequent round of re-approvals....

In any case, I deem the code to be fit, unless either of the comments below persuades you otherwise.

File: lib/pbench/test/unit/agent/test_utils.py
Comment (portante) [reply lib/pbench/test/unit/agent/test_utils.py:43]: `value` is the exception object that was raised.  Go figure.

File: lib/pbench/test/unit/agent/test_utils.py
Comment (portante) [reply lib/pbench/test/unit/agent/test_utils.py:35]: I don't think so.  Using the `OurServer.def_port` class variable is fine because to avoid further perpetuating a constant, but `localhost` is so common that it doesn't gain us much making it dynamic.

**Review by ndokos — APPROVED** [review]

**Review by webbnh — APPROVED** [review]
GitHub sez nothing has changed since my last review, so it must be good to go!

**Pull request merged**
